File: shared-domain/src/commonMain/kotlin/com/dualmusic/domain/realtime/RealtimeContract.kt
package com.dualmusic.domain.realtime

import com.dualmusic.domain.model.EventContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Contrat temps réel Dual Music — **Socket.IO** (et NON Supabase).
 *
 * Référence unique partagée iOS/Android : namespaces, nommage des rooms, catalogue
 * d'événements et formes de payloads. Les clients `core-realtime` (un par plateforme)
 * implémentent ce contrat ; toute évolution se fait ICI d'abord.
 *
 * Modèle backend :
 *  - 3 namespaces : `/chat`, `/live`, `/notifications`.
 *  - handshake authentifié par JWT (le token est fourni à la connexion).
 *  - on rejoint une room en émettant `join { type, id }` ; le nom interne = `type:id`.
 *  - le serveur pousse des événements dans la room (voir [RealtimeEvent]) et à
 *    l'utilisateur (événements `tx:*`).
 */
object Realtime {

    /** Namespaces Socket.IO exposés par le backend. */
    object Namespace {
        const val CHAT = "/chat"
        const val LIVE = "/live"
        const val NOTIFICATIONS = "/notifications"
    }

    /**
     * Types de room joignables (correspondent à [EventContext] + `leaderboard`).
     * La valeur wire est utilisée dans le payload `join { type, id }`.
     */
    enum class RoomType(val wire: String) {
        DUEL("duel"),
        CONCERT("concert"),
        COMPETITION("competition"),
        LIVE("live"),
        LEADERBOARD("leaderboard");

        companion object {
            /** Convertit un [EventContext] métier en [RoomType] realtime. */
            fun from(context: EventContext): RoomType = when (context) {
                EventContext.DUEL -> DUEL
                EventContext.CONCERT -> CONCERT
                EventContext.COMPETITION -> COMPETITION
                EventContext.LIVE -> LIVE
            }
        }
    }

    /** Nom interne d'une room : `type:id` (identique au backend `roomName`). */
    fun roomName(type: RoomType, id: String): String = "${type.wire}:$id"

    /** Événement client → serveur pour rejoindre/quitter une room. */
    const val EVENT_JOIN = "join"
    const val EVENT_LEAVE = "leave"

    /**
     * Catalogue des événements serveur → client.
     * Documenté avec la room/namespace d'origine pour brancher les bons listeners.
     */
    object RealtimeEvent {
        /** `/live` · room duel — un vote payant vient d'être enregistré. Payload [VotePayload]. */
        const val VOTE = "vote"

        /** `/live` · room event — un cadeau a été envoyé dans l'événement. Payload [GiftPayload]. */
        const val GIFT = "gift"

        /** par-utilisateur — le destinataire d'un cadeau est notifié (toast). Payload [TxGiftPayload]. */
        const val TX_GIFT = "tx:gift"

        /** par-utilisateur — un crédit (recharge/gain) a été porté au solde. Payload [TxCreditPayload]. */
        const val TX_CREDIT = "tx:credit"

        /** par-utilisateur — un retrait a changé d'état. Payload [TxWithdrawalPayload]. */
        const val TX_WITHDRAWAL = "tx:withdrawal"

        /** `/live` · room event — changement de statut (upcoming/live/ended/winner). Payload [StatusPayload]. */
        const val STATUS = "status"

        /** `/live` · room duel — minuteur (start/stop, cible). Payload [TimerPayload]. */
        const val TIMER = "timer"

        /** `/live` · room event — présence (viewers). Payload [PresencePayload]. */
        const val PRESENCE = "presence"

        /** `/live` · room compétition — désignation du performeur courant + durée. Payload [PerformerPayload]. */
        const val PERFORMER = "performer"

        /** `/live` · room compétition — focus caméra imposé par le manager. Payload [FocusPayload]. */
        const val FOCUS = "focus"

        /** `/live` · room event — diffusion d'une pub sponsor (start/stop). Payload [SponsorAdPayload]. */
        const val SPONSOR_AD = "sponsor:ad"

        /** `/chat` · room event — nouveau message de chat. Payload [ChatMessagePayload]. */
        const val CHAT_MESSAGE = "message"

        /** `/notifications` — nouvelle notification in-app. Payload libre (voir feature notifications). */
        const val NOTIFICATION = "notification"

        /**
         * `/live` · room event — un réglage a changé (au minimum `chat_enabled` ; live envoie
         * aussi `allows_dedications`/`allow_guests`). Payload [EventSettingsPayload].
         */
        const val SETTINGS = "settings"

        /** `/live` · room event — l'hôte a désigné un spectateur modérateur. Payload [EventModeratorPayload]. */
        const val MODERATOR_APPOINTED = "moderator:appointed"

        /** `/live` · room event — l'hôte a révoqué un modérateur désigné. Payload [EventModeratorPayload]. */
        const val MODERATOR_REVOKED = "moderator:revoked"
    }
}

/** Payload de l'événement `join`/`leave`. */
@Serializable
data class JoinPayload(val type: String, val id: String)

/** `vote` — nouveau vote sur un duel. */
@Serializable
data class VotePayload(
    @SerialName("duel_id") val duelId: String,
    @SerialName("artist_id") val artistId: String,
    val amount: Double,
    @SerialName("user_id") val userId: String? = null,
)

/** `gift` — cadeau envoyé dans une room d'événement. */
@Serializable
data class GiftPayload(
    @SerialName("to_user_id") val toUserId: String? = null,
    @SerialName("from_user_id") val fromUserId: String? = null,
    val value: Double = 0.0,
    @SerialName("gift_id") val giftId: String? = null,
    @SerialName("gift_name") val giftName: String? = null,
    @SerialName("gift_image") val giftImage: String? = null,
)

/** Vidéo publicitaire sponsor diffusable dans un événement. */
@Serializable
data class SponsorAdVideo(
    val id: String,
    @SerialName("video_url") val videoUrl: String,
    val title: String = "",
    @SerialName("duration_seconds") val durationSeconds: Int = 0,
    @SerialName("play_count") val playCount: Int = 0,
)

/** `sponsor:ad` — début/fin de diffusion d'une pub dans la room (start/stop). */
@Serializable
data class SponsorAdPayload(
    val action: String, // "start" | "stop"
    @SerialName("play_id") val playId: String? = null,
    val ad: SponsorAdVideo? = null,
)

/** Réponse de `POST /sponsors/ads/play` : l'enregistrement de diffusion créé. */
@Serializable
data class SponsorAdPlay(val id: String)

/** `tx:gift` — notification au destinataire d'un cadeau. */
@Serializable
data class TxGiftPayload(val amount: Double)

/** `tx:credit` — crédit porté au solde de l'utilisateur. */
@Serializable
data class TxCreditPayload(val amount: Double, val currency: String? = null)

/** `tx:withdrawal` — mise à jour d'un retrait. */
@Serializable
data class TxWithdrawalPayload(
    val amount: Double,
    val status: String,
)

/** `status` — changement d'état d'un duel/live/concert/compétition. */
@Serializable
data class StatusPayload(
    @SerialName("duel_id") val duelId: String? = null,
    @SerialName("live_id") val liveId: String? = null,
    val status: String,
    @SerialName("winner_id") val winnerId: String? = null,
    @SerialName("room_id") val roomId: String? = null,
)

/** `timer` — minuteur d'un duel (arrivants tardifs inclus). */
@Serializable
data class TimerPayload(
    @SerialName("duel_id") val duelId: String? = null,
    @SerialName("ends_at") val endsAt: String? = null,
    @SerialName("target_id") val targetId: String? = null,
)

/** `presence` — compteur de spectateurs d'un live. Le backend émet `{ room: "<type>:<id>", count }`. */
@Serializable
data class PresencePayload(
    @SerialName("live_id") val liveId: String? = null,
    /** Nom de room (`live:<id>`) — sert à identifier le live en écoute multi-room. */
    val room: String? = null,
    val count: Int,
)

/**
 * `performer` — le manager désigne le candidat qui performe (ou `null` pour arrêter) + la durée
 * de son slot. Le backend émet en camelCase. Pas de `startedAt` : le client calcule la fin à
 * la réception (`maintenant + durationSec`), l'event arrivant quasi immédiatement.
 */
@Serializable
data class PerformerPayload(
    val competitionId: String? = null,
    val performerId: String? = null,
    val durationSec: Int = 0,
)

/**
 * `focus` — le manager impose (ou libère avec `null`) la caméra mise en avant pour TOUS les
 * spectateurs. `participantId` = identité LiveKit du publieur épinglé (= userId). camelCase.
 */
@Serializable
data class FocusPayload(
    val competitionId: String? = null,
    val participantId: String? = null,
)

/**
 * `settings` — au minimum `chatEnabled` (déduit de `chat_enabled`, présent sur les 4 types
 * d'évènement) ; live envoie en plus `allowsDedications`/`dedicationMinPriceCredits`/`allowGuests`.
 * Un seul champ id est renseigné selon le type d'évènement.
 */
@Serializable
data class EventSettingsPayload(
    @SerialName("duel_id") val duelId: String? = null,
    @SerialName("live_id") val liveId: String? = null,
    @SerialName("concert_id") val concertId: String? = null,
    @SerialName("competition_id") val competitionId: String? = null,
    @SerialName("chat_enabled") val chatEnabled: Boolean? = null,
    @SerialName("allows_dedications") val allowsDedications: Boolean? = null,
    @SerialName("dedication_min_price_credits") val dedicationMinPriceCredits: Double? = null,
    @SerialName("allow_guests") val allowGuests: Boolean? = null,
)

/** `moderator:appointed` / `moderator:revoked` — un modérateur d'évènement a été nommé/révoqué. */
@Serializable
data class EventModeratorPayload(
    @SerialName("event_type") val eventType: String,
    @SerialName("event_id") val eventId: String,
    @SerialName("user_id") val userId: String,
)

/** `message` — message de chat d'une room. */
@Serializable
data class ChatMessagePayload(
    val id: String? = null,
    @SerialName("user_id") val userId: String,
    // Le backend utilise la clé `message` (colonne DB), pas `content`.
    @SerialName("message") val content: String,
    @SerialName("created_at") val createdAt: String? = null,
    // Le backend diffuse l'auteur sous la clé `author` (chat.service.js) — pas `user`.
    @SerialName("author") val user: com.dualmusic.domain.model.DisplayProfile? = null,
    // Réponse à un message parent (chat en fil) — présent quand c'est une réponse.
    @SerialName("parent_id") val parentId: String? = null,
)

File: ios/Packages/DualMusicKit/Sources/DomainModels/RealtimeContract.swift
import Foundation

/*
 * Contrat temps réel Dual Music — **Socket.IO** (et NON Supabase).
 *
 * Miroir de `shared-domain/realtime/RealtimeContract.kt` : namespaces, nommage des rooms,
 * catalogue d'événements et formes de payloads. `CoreRealtime` implémente ce contrat ;
 * toute évolution se fait d'abord dans `shared-domain`, puis ici et côté Android.
 *
 * Modèle backend :
 *  - 3 namespaces : `/chat`, `/live`, `/notifications` ;
 *  - handshake authentifié par JWT (fourni à la connexion) ;
 *  - on rejoint une room en émettant `join { type, id }` ; nom interne = `type:id` ;
 *  - le serveur pousse dans la room (voir ``Realtime/Event``) et à l'utilisateur (`tx:*`).
 */
public enum Realtime {

    /// Namespaces Socket.IO exposés par le backend.
    public enum Namespace: String, Sendable, Hashable {
        case chat = "/chat"
        case live = "/live"
        case notifications = "/notifications"
    }

    /// Types de room joignables (correspondent à ``EventContext`` + `leaderboard`).
    /// La valeur brute est celle utilisée dans le payload `join { type, id }`.
    public enum RoomType: String, Sendable {
        case duel, concert, competition, live, leaderboard

        /// Convertit un contexte métier en type de room temps réel.
        /// - Parameter context: contexte d'événement.
        public static func from(_ context: EventContext) -> RoomType {
            switch context {
            case .duel: return .duel
            case .concert: return .concert
            case .competition: return .competition
            case .live: return .live
            }
        }
    }

    /// Nom interne d'une room : `type:id` (identique au `roomName` backend).
    public static func roomName(_ type: RoomType, _ id: String) -> String { "\(type.rawValue):\(id)" }

    /// Événement client → serveur pour rejoindre une room.
    public static let eventJoin = "join"
    /// Événement client → serveur pour quitter une room.
    public static let eventLeave = "leave"

    /// Catalogue des événements serveur → client.
    public enum Event {
        /// `/live` · room duel — un vote payant vient d'être enregistré. Payload ``VotePayload``.
        public static let vote = "vote"
        /// `/live` · room event — un cadeau a été envoyé. Payload ``GiftPayload``.
        public static let gift = "gift"
        /// par-utilisateur — le destinataire d'un cadeau est notifié. Payload ``TxGiftPayload``.
        public static let txGift = "tx:gift"
        /// par-utilisateur — un crédit a été porté au solde. Payload ``TxCreditPayload``.
        public static let txCredit = "tx:credit"
        /// par-utilisateur — un retrait a changé d'état. Payload ``TxWithdrawalPayload``.
        public static let txWithdrawal = "tx:withdrawal"
        /// `/live` · room event — changement de statut. Payload ``StatusPayload``.
        public static let status = "status"
        /// `/live` · room duel — minuteur (start/stop, cible). Payload ``TimerPayload``.
        public static let timer = "timer"
        /// `/live` · room event — présence (viewers). Payload ``PresencePayload``.
        public static let presence = "presence"
        /// `/chat` · room event — nouveau message. Payload ``ChatMessagePayload``.
        public static let chatMessage = "message"
        /// `/notifications` — nouvelle notification in-app. Payload ``AppNotification``.
        public static let notification = "notification"
        /// `/live` · room event — réglages modifiés par l'hôte (dédicaces/invités/chat).
        /// Payload ``LiveSettingsPayload``.
        public static let settings = "settings"
        /// `/live` · room live — nouvelle demande de dédicace. Payload ``DedicationEventPayload``.
        public static let dedicationNew = "dedication:new"
        /// `/live` · room live — dédicace acceptée/rejetée/livrée. Payload
        /// ``DedicationEventPayload``.
        public static let dedicationUpdate = "dedication:update"
        /// `/live` · room live — nouvelle demande d'invité (« lever la main »). Payload
        /// ``JoinEventPayload``.
        public static let joinNew = "join:new"
        /// `/live` · room live — demande d'invité acceptée/rejetée/terminée/annulée. Payload
        /// ``JoinEventPayload``.
        public static let joinUpdate = "join:update"
    }
}

// MARK: - Payloads

/// `vote` — nouveau vote sur un duel.
public struct VotePayload: Decodable, Sendable {
    public let duelId: String
    public let artistId: String
    public let amount: Double
    public let userId: String?

    enum CodingKeys: String, CodingKey {
        case duelId = "duel_id"
        case artistId = "artist_id"
        case amount
        case userId = "user_id"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        duelId = c.val(String.self, .duelId, "")
        artistId = c.val(String.self, .artistId, "")
        amount = c.amount(.amount)
        userId = c.opt(String.self, .userId)
    }
}

/// `gift` — cadeau envoyé dans une room d'événement.
public struct GiftPayload: Decodable, Sendable {
    public let toUserId: String?
    public let fromUserId: String?
    public let value: Double
    public let giftId: String?
    public let giftName: String?
    public let giftImage: String?

    enum CodingKeys: String, CodingKey {
        case toUserId = "to_user_id"
        case fromUserId = "from_user_id"
        case value
        case giftId = "gift_id"
        case giftName = "gift_name"
        case giftImage = "gift_image"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        toUserId = c.opt(String.self, .toUserId)
        fromUserId = c.opt(String.self, .fromUserId)
        value = c.amount(.value)
        giftId = c.opt(String.self, .giftId)
        giftName = c.opt(String.self, .giftName)
        giftImage = c.opt(String.self, .giftImage)
    }
}

/// `tx:gift` — notification au destinataire d'un cadeau.
public struct TxGiftPayload: Decodable, Sendable {
    public let amount: Double

    enum CodingKeys: String, CodingKey { case amount }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        amount = c.amount(.amount)
    }
}

/// `tx:credit` — crédit porté au solde de l'utilisateur.
public struct TxCreditPayload: Decodable, Sendable {
    public let amount: Double
    public let currency: String?

    enum CodingKeys: String, CodingKey { case amount, currency }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        amount = c.amount(.amount)
        currency = c.opt(String.self, .currency)
    }
}

/// `tx:withdrawal` — mise à jour d'un retrait.
public struct TxWithdrawalPayload: Decodable, Sendable {
    public let amount: Double
    public let status: String

    enum CodingKeys: String, CodingKey { case amount, status }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        amount = c.amount(.amount)
        status = c.val(String.self, .status, "")
    }
}

/// `status` — changement d'état d'un duel/live/concert/compétition.
public struct StatusPayload: Decodable, Sendable {
    public let duelId: String?
    public let liveId: String?
    public let status: String
    public let winnerId: String?
    public let roomId: String?

    enum CodingKeys: String, CodingKey {
        case duelId = "duel_id"
        case liveId = "live_id"
        case status
        case winnerId = "winner_id"
        case roomId = "room_id"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        duelId = c.opt(String.self, .duelId)
        liveId = c.opt(String.self, .liveId)
        status = c.val(String.self, .status, "")
        winnerId = c.opt(String.self, .winnerId)
        roomId = c.opt(String.self, .roomId)
    }
}

/// `timer` — minuteur d'un duel (arrivants tardifs inclus).
public struct TimerPayload: Decodable, Sendable {
    public let duelId: String?
    public let endsAt: String?
    public let targetId: String?

    enum CodingKeys: String, CodingKey {
        case duelId = "duel_id"
        case endsAt = "ends_at"
        case targetId = "target_id"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        duelId = c.opt(String.self, .duelId)
        endsAt = c.opt(String.self, .endsAt)
        targetId = c.opt(String.self, .targetId)
    }
}

/// `presence` — compteur de spectateurs d'un live.
public struct PresencePayload: Decodable, Sendable {
    public let liveId: String?
    public let count: Int

    enum CodingKeys: String, CodingKey {
        case liveId = "live_id"
        case count
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        liveId = c.opt(String.self, .liveId)
        count = c.int(.count)
    }
}

/// `message` — message de chat d'une room.
public struct ChatMessagePayload: Decodable, Sendable {
    public let id: String?
    public let userId: String
    public let content: String
    public let createdAt: String?
    public let user: DisplayProfile?

    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case content
        case createdAt = "created_at"
        case user
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.opt(String.self, .id)
        userId = c.val(String.self, .userId, "")
        content = c.val(String.self, .content, "")
        createdAt = c.opt(String.self, .createdAt)
        user = c.opt(DisplayProfile.self, .user)
    }
}

/// `settings` — réglages d'un live modifiés par l'hôte, en direct (dédicaces/invités/chat).
/// `dedicationMinPriceCredits` absent = pas de surcharge sur CE live (garder le prix déjà
/// chargé), à ne pas confondre avec 0.
public struct LiveSettingsPayload: Decodable, Sendable {
    public let allowsDedications: Bool
    public let dedicationMinPriceCredits: Double?
    public let allowGuests: Bool
    public let chatEnabled: Bool?

    enum CodingKeys: String, CodingKey {
        case allowsDedications = "allows_dedications"
        case dedicationMinPriceCredits = "dedication_min_price_credits"
        case allowGuests = "allow_guests"
        case chatEnabled = "chat_enabled"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        allowsDedications = c.bool(.allowsDedications, true)
        dedicationMinPriceCredits = c.amountIfPresent(.dedicationMinPriceCredits)
        allowGuests = c.bool(.allowGuests, true)
        chatEnabled = c.opt(Bool.self, .chatEnabled)
    }
}

/// `dedication:new` / `dedication:update` — même room que `join:live`. Sert seulement à
/// détecter "est-ce que ça me concerne" (fan) ou "dois-je recharger" (hôte) : l'état fait
/// toujours l'objet d'un rechargement REST complet, jamais appliqué depuis ce payload.
public struct DedicationEventPayload: Decodable, Sendable {
    public let dedicationId: String?
    public let fanId: String?
    public let status: String?
    public let priceCredits: Double?

    enum CodingKeys: String, CodingKey {
        case dedicationId = "dedication_id"
        case fanId = "fan_id"
        case status
        case priceCredits = "price_credits"
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        dedicationId = c.opt(String.self, .dedicationId)
        fanId = c.opt(String.self, .fanId)
        status = c.opt(String.self, .status)
        priceCredits = c.amountIfPresent(.priceCredits)
    }
}

/// `join:new` / `join:update` — même room que `join:live` (dédicaces). Sert seulement à
/// détecter "dois-je recharger" (hôte, ou tout le monde pour recalculer les abonnements aux
/// rooms d'invités) et "ma propre demande a changé d'état" (spectateur concerné) : l'état
/// fait toujours l'objet d'un rechargement REST complet.
public struct JoinEventPayload: Decodable, Sendable {
    public let liveId: String?
    public let userId: String?
    public let status: String?

    enum CodingKeys: String, CodingKey {
        case liveId = "live_id"
        case userId = "user_id"
        case status
    }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        liveId = c.opt(String.self, .liveId)
        userId = c.opt(String.self, .userId)
        status = c.opt(String.self, .status)
    }
}
